### PR TITLE
feat: Upgrade to Terraform 0.12

### PR DESCRIPTION
**Pull request by syphernl**
## what
* Synced up repo config with that of other CP repos (chatops, releasing, go, actions etc)
* Update any version pinning
* Added a type to all variables
* Modified `ordered_cache_behavior` to match that of other Cloudfront modules (**breaking change**: `cache_behavior` list renamed to `ordered_cache`)
* (Locally verified) no terraform init errors using cli

## why
* Support Terraform 0.12
* Provide next steps for adding support for Terraform 0.13

## notes
* closes #27 
* closes #29 
* closes #34 
* closes #25

### CODE REVIEW

File: examples/basic/variables.tf [hunk 1]
```diff
@@ -0,0 +1,281 @@
+variable "region" {
```
Comment (aknysh) [inline examples/basic/variables.tf:1]: @syphernl thanks for the PR!
Please move the example from `examples/basic` to `examples/complete` folder - we use it for all our examples that get tested on AWS by terratest in all modules

File: main.tf [hunk 2]
```diff
@@ -1,123 +1,167 @@
 module "origin_label" {
-  source     = "git::https://github.com/cloudposse/terraform-null-label.git?ref=tags/0.3.7"
-  namespace  = "${var.namespace}"
-  stage      = "${var.stage}"
-  name       = "${var.name}"
-  delimiter  = "${var.delimiter}"
-  attributes = ["${compact(concat(var.attributes, list("origin")))}"]
-  tags       = "${var.tags}"
+  source     = "git::https://github.com/cloudposse/terraform-null-label.git?ref=tags/0.17.0"
+  namespace  = var.namespace
```
Comment (aknysh) [inline main.tf:3]: please add `var.environment`, we are adding it to all our modules now.
See https://github.com/cloudposse/terraform-aws-eks-fargate-profile/blob/master/examples/complete/variables.tf#L21
https://github.com/cloudposse/terraform-aws-eks-fargate-profile/blob/master/variables.tf#L13
https://github.com/cloudposse/terraform-aws-eks-fargate-profile/blob/master/main.tf#L13

File: main.tf [hunk 3]
```diff
@@ -1,123 +1,167 @@
...
-  source     = "git::https://github.com/cloudposse/terraform-null-label.git?ref=tags/0.3.7"
-  namespace  = "${var.namespace}"
-  stage      = "${var.stage}"
-  name       = "${var.name}"
-  attributes = "${var.attributes}"
-  delimiter  = "${var.delimiter}"
-  tags       = "${var.tags}"
+  source     = "git::https://github.com/cloudposse/terraform-null-label.git?ref=tags/0.17.0"
+  namespace  = var.namespace
+  stage      = var.stage
```
Comment (aknysh) [inline main.tf:32]: please add `var.environment`

File: test/src/examples_basic_test.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+		// Variables to pass to our Terraform code using -var-file options
+		VarFiles: []string{"fixtures.us-east-2.tfvars"},
+	}
+
+	// At the end of the test, run `terraform destroy` to clean up any resources that were created
+	defer terraform.Destroy(t, terraformOptions)
+
+	// This will run `terraform init` and `terraform apply` and fail the test if there are any errors
+	terraform.InitAndApply(t, terraformOptions)
+}
```
Comment (aknysh) [inline test/src/examples_basic_test.go:41]: would be nice to add some actual testing of the outputs here, e.g. https://github.com/cloudposse/terraform-aws-eks-fargate-profile/blob/master/test/src/examples_complete_test.go#L96

File: variables.tf [hunk 5]
```diff
@@ -1,243 +1,314 @@
...
 variable "web_acl_id" {
-  description = "(Optional) - Web ACL ID that can be attached to the Cloudfront distribution"
+  type        = string
+  description = "ID of the AWS WAF web ACL that is associated with the distribution"
   default     = ""
 }
 
 variable "origin_domain_name" {
+  type        = string
   description = "(Required) - The DNS domain name of your custom origin (e.g. website)"
```
Comment (aknysh) [inline variables.tf:82]:
```suggestion
  description = "The DNS domain name of your custom origin (e.g. website)"
```

File: variables.tf [hunk 6]
```diff
@@ -1,243 +1,314 @@
...
-  description = "(Optional) - An optional element that causes CloudFront to request your content from a directory in your Amazon S3 bucket or your custom origin"
+  # http://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/distribution-web-values-specify.html#DownloadDistValuesOriginPath
+  type        = string
+  description = "An optional element that causes CloudFront to request your content from a directory in your Amazon S3 bucket or your custom origin. It must begin with a /. Do not add a / at the end of the path."
   default     = ""
 }
 
 variable "origin_http_port" {
+  type        = number
   description = "(Required) - The HTTP port the custom origin listens on"
```
Comment (aknysh) [inline variables.tf:95]:
```suggestion
  description = "The HTTP port the custom origin listens on"
```

File: variables.tf [hunk 7]
```diff
@@ -1,243 +1,314 @@
...
   default     = "80"
 }
 
 variable "origin_https_port" {
+  type        = number
   description = "(Required) - The HTTPS port the custom origin listens on"
```
Comment (aknysh) [inline variables.tf:101]:
```suggestion
  description = "The HTTPS port the custom origin listens on"
```

File: variables.tf [hunk 8]
```diff
@@ -1,243 +1,314 @@
...
-  default     = "443"
+  default     = 443
 }
 
 variable "origin_protocol_policy" {
+  type        = string
   description = "(Required) - The origin protocol policy to apply to your origin. One of http-only, https-only, or match-viewer"
```
Comment (aknysh) [inline variables.tf:107]:
```suggestion
  description = "The origin protocol policy to apply to your origin. One of http-only, https-only, or match-viewer"
```

File: variables.tf [hunk 9]
```diff
@@ -1,243 +1,314 @@
...
   default     = "match-viewer"
 }
 
 variable "origin_ssl_protocols" {
   description = "(Required) - The SSL/TLS protocols that you want CloudFront to use when communicating with your origin over HTTPS"
```
Comment (aknysh) [inline variables.tf:112]:
```suggestion
  description = "The SSL/TLS protocols that you want CloudFront to use when communicating with your origin over HTTPS"
```

File: variables.tf [hunk 10]
```diff
@@ -1,243 +1,314 @@
...
-  type        = "list"
+  type        = list(string)
   default     = ["TLSv1", "TLSv1.1", "TLSv1.2"]
 }
 
 variable "origin_keepalive_timeout" {
+  type        = number
   description = "(Optional) The Custom KeepAlive timeout, in seconds. By default, AWS enforces a limit of 60. But you can request an increase."
```
Comment (aknysh) [inline variables.tf:119]:
```suggestion
  description = "The Custom KeepAlive timeout, in seconds. By default, AWS enforces a limit of 60. But you can request an increase."
```

File: variables.tf [hunk 11]
```diff
@@ -1,243 +1,314 @@
...
-  default     = "60"
+  default     = 60
 }
 
 variable "origin_read_timeout" {
+  type        = number
   description = "(Optional) The Custom Read timeout, in seconds. By default, AWS enforces a limit of 60. But you can request an increase."
```
Comment (aknysh) [inline variables.tf:125]:
```suggestion
  description = "The Custom Read timeout, in seconds. By default, AWS enforces a limit of 60. But you can request an increase."
```

File: variables.tf [hunk 12]
```diff
@@ -1,243 +1,314 @@
...
-  default     = "60"
+  default     = 60
 }
 
 variable "compress" {
+  type        = bool
   description = "(Optional) Whether you want CloudFront to automatically compress content for web requests that include Accept-Encoding: gzip in the request header (default: false)"
```
Comment (aknysh) [inline variables.tf:131]:
```suggestion
  description = "Whether you want CloudFront to automatically compress content for web requests that include Accept-Encoding: gzip in the request header (default: false)"
```

File: variables.tf [hunk 13]
```diff
@@ -1,243 +1,314 @@
...
 
 variable "price_class" {
+  type        = string
   default     = "PriceClass_100"
   description = "Price class for this distribution: `PriceClass_All`, `PriceClass_200`, `PriceClass_100`"
 }
 
 variable "viewer_minimum_protocol_version" {
+  type        = string
   description = "(Optional) The minimum version of the SSL protocol that you want CloudFront to use for HTTPS connections."
```
Comment (aknysh) [inline variables.tf:215]:
```suggestion
  description = "The minimum version of the SSL protocol that you want CloudFront to use for HTTPS connections."
```

File: test/src/examples_basic_test.go (hunk 4 above)
Comment (syphernl) [reply test/src/examples_basic_test.go:51]: I have added a few tests of the outputs.

**Comment by Gowiem** [comment]
@aknysh Since these are new tests / chatops is just being installed I don't think we can use our actions to run this project like @syphernl is trying. What is our process for getting this merged considering we can't run it here? Pull it down and provide manual validation that these tests are passing?

**Comment by aknysh** [comment]
> @aknysh Since these are new tests / chatops is just being installed I don't think we can use our actions to run this project like @syphernl is trying. What is our process for getting this merged considering we can't run it here? Pull it down and provide manual validation that these tests are passing?

we need to open another PR first and add all the actions in there, then merge to master.
I'll do it now

**Comment by aknysh** [comment]
@syphernl please rebase

**Comment by syphernl** [comment]
@aknysh Done!

**Comment by osterman** [comment]
The terratest run is failing:

image

File: test/src/examples_complete_test.go [hunk 14]
```diff
@@ -0,0 +1,51 @@
+package test
+
+import (
+	"math/rand"
+	"testing"
+	"time"
+
+	"github.com/gruntwork-io/terratest/modules/terraform"
+)
```
Comment (aknysh) [inline test/src/examples_complete_test.go:9]:
```
github.com/stretchr/testify/assert"
```

This needs to be added.

Also please execute the following commands:

```
cd test/src/
go mod init github.com/cloudposse/$(cd ../../ && basename `pwd`)
go mod tidy
git rm Gopkg.*
git add go.*
```

**Review by aknysh — COMMENTED** [review]
please see comments

**Comment by syphernl** [comment]
> please see comments

Done!

**Comment by jamengual** [comment]
@syphernl please look at the tests, there seems to be failing due to a missing dns zone.

**Pull request closed without merging**